ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java

```diff
@@ -877,6 +939,61 @@ MessageFormatFlags getOperationFlag() {
           : MessageFormatFlags.All;
     }
 
+    @Override
+    protected boolean maybeProcessDecryptCallbackAndCompleteOperation() {
+      if (decryptionRequired.get() && decryptCallbackResultInfo.resultAvailable) {
+        // in case of Metadata blob, only user-metadata needs decryption if the blob is encrypted
+        if (blobType == BlobType.MetadataBlob) {
+          if (decryptCallbackResultInfo.exception == null) {
+            logger.trace("Processing stored decryption callback result for Metadata blob {}", blobId);
+            chunkIdIterator = keys.listIterator();
+            numChunksTotal = keys.size();
+            dataChunks = new GetChunk[Math.min(keys.size(), NonBlockingRouter.MAX_IN_MEM_CHUNKS)];
+            for (int i = 0; i < dataChunks.length; i++) {
+              dataChunks[i] = new GetChunk(chunkIdIterator.nextIndex(), (BlobId) chunkIdIterator.next());
+            }
+            blobInfo = new BlobInfo(blobInfo.getBlobProperties(),
+                decryptCallbackResultInfo.result.getDecryptedUserMetadata().array());
+            decryptionStatusTracker.setSucceeded();
+          } else {
+            logger.trace("Decryption job callback invoked with exception {} for Metadata blob {} ",
+                decryptCallbackResultInfo.exception, blobId);
+            setOperationException(
+                new RouterException("Exception thrown on decrypting content for Metadata blob " + blobId,
+                    RouterErrorCode.UnexpectedInternalError));
+            decryptionStatusTracker.setFailed();
+          }
+          logger.trace("BlobContent available to process for Metadata blob {}", blobId);
+        } else {
+          // Incase of simple blobs, user-metadata may or may not be passed into decryption job based on GetOptions flag.
```

Comment:
These comments are confusing. Why talk about `GetBlobInfo` here? You shouldn't have to talk about `GetOptions` either. Basically, deserialization of `BlobAll` or deserialization of just `BlobData` will tell you whether the data is encrypted or not.

Reply:
sorry. this comment needs to go in two lines later. just before setting the right BlobInfo. Let me know if it still doesn't make sense.